keps/sig-cluster-lifecycle/addons/20191013-install-addons-via-kubeadm.md

```diff
@@ -0,0 +1,450 @@
+
+The AddonInstaller feature composes with kubeadm's global `--dry-run` flag:
+- on `init`, the intended addons (and perhaps manifests) can simply be printed.
+- on `upgrade diff` and `upgrade apply`, the expected difference relative to the cluster can be output.
+  `kubectl` already does this currently.
+
+The difference in behavior between a dryRun init and upgrade is caused by the presence of a working KUBECONFIG and control-plane.
+
+Air-gap support may be accomplished through the mirroring of git repositories or images containing the kustomize dir.
+Mirroring git repositories, cloning git repositories to a local filesystem, and hosting OCI images would seem to all be a reasonable capabilities for a user in an air-gapped environment. The API allows a user to reference addon sources for all of these use-cases.
+
+#### Links
+
+@fabriziopandini began a design doc for kubeadm integration:
+- https://docs.google.com/document/d/1FZA873LBXf-aG4UULOk96MhT8xm_nPHq5Vu0z61S8pk
+
+More generic User Stories have been collected by @dholbach:
+- doc: https://docs.google.com/document/d/17NH4xcFdeh4NVcIBjMQNK2P27j9xrsm4sCFTXdZqopg
+- sheet: https://docs.google.com/spreadsheets/d/1Np0aOQYyiRqRQYOUoMcUwP_2Tt9b39HUwJXQIgW9rYA
+
+### Implementation Details/Notes/Constraints
+
+We intend for the installer library and API to live in `sigs.k8s.io/addon-operators/installer`.
+Core kustomize addons and images may be hosted in a number of git repositories or image registries.
+
+Internally, the addon-installer library execs `kubectl -k`.
+This introduces a binary and PATH dependency for `kubeadm` when using this feature.
+Furthermore, certain sensible usage of this feature may depend on `git` by proxy of kustomize.
+While `kubectl` may commonly be available in the cluster-installation environment, `git` may not be available.
+ex: The current kinder images do not have `git`.
+
+It may be more appropriate to exec `kustomize` directly or vendor it in for manifest generation so that dryRun and validation of the kustomize addons will work in the absence of a working kube-apiserver. (This is important before `kubeadm init`)
+
+The kubeconfig may be passed to the library as a filepath.
+This works for kubeadm and means less work when exec'ing `kubectl` for the apply logic.
+Should we also support a formal internal type for this?
+
+A more mature addon-installer library may:
+- Vendor kustomize for manifest generation.
+- Use an apiClient type for applying the manifests.
+  ( This may be compatible with some shareable client type supporting dryRun? )
+- Have an go-native git clone implementation
+
+Kustomize currently supports constructing manifests from local directories and git.
+OCI image support in kustomize is currently unimplemented.
+A POC of some of this mechanism is implemented here: https://github.com/ecordell/kpg
+Implementation of this feature could be delegated to another KEP.
+
+### Risks and Mitigations
```

Comment:
I think it is missing a risk about impacts on kubeadm users / third part tools using kubeadm